lms/validation/_lis_result_sourcedid.py

```diff
@@ -0,0 +1,58 @@
+"""Schema for validating params required for LIS Outcomes."""
+import marshmallow
+
+from lms.validation._helpers import PyramidRequestSchema
+from lms.values import LISResultSourcedId
+
+
+__all__ = "LISResultSourcedIdSchema"
+
+
+class LISResultSourcedIdSchema(PyramidRequestSchema):
+    """
+    Schema for LIS Result/LMS Outcomes params.
+
+    Validates params for LMS Outcomes metadata. Usage via webargs::
+
+        >>> from webargs.pyramidparser import parser
+        >>>
+        >>> schema = LISResultSourcedIdSchema(request)
+        >>> parsed_params = parser.parse(schema, request)
+
+    Or to verify the request and get an :class:`~lms.values.LISResultSourcedId`
+    from the request's params::
+
+        >>> schema = LISResultSourcedIdSchema(request)
+        >>> schema.lis_result_sourcedid()
+        LISResultSourcedId(lis_result_sourcedid='...', ...)
+    """
+
+    lis_result_sourcedid = marshmallow.fields.Str(required=True)
+    lis_outcome_service_url = marshmallow.fields.Str(required=True)
+    context_id = marshmallow.fields.Str(required=True)
+    resource_link_id = marshmallow.fields.Str(required=True)
+    tool_consumer_info_product_family_code = marshmallow.fields.Str(required=False)
+
+    def lis_result_sourcedid_info(self):
+        """
+        Return an :class:`~lms.values.LISResultSourcedId`.
+
+        Return an :class:`~lms.values.LISResultSourcedId` generated from the
+        request's validated launch params, along with the passed ``HUser`` and
+        ``LTIUser`` ``NamedTuples``.
+
+        :raise ValidationError: if the request isn't a valid LIS Outcome launch
+        :rtype: LISResultSourcedId
+        """
+
+        kwargs = self.parse()
+
+        return LISResultSourcedId(
+            lis_result_sourcedid=kwargs["lis_result_sourcedid"],
+            lis_outcome_service_url=kwargs["lis_outcome_service_url"],
+            context_id=kwargs["context_id"],
+            resource_link_id=kwargs["resource_link_id"],
+            tool_consumer_info_product_family_code=kwargs[
+                "tool_consumer_info_product_family_code"
+            ],
```

Comment:
Does this crash if the optional `"tool_consumer_info_product_family_code"` is missing?

Reply:
This test should test that (if it doesn't do it right, let me know!) https://github.com/hypothesis/lms/pull/974/files#diff-fadd7e77a433f71faaa313f7bf17e438R22

Reply:
That test calls `parse()`, not `lis_result_sourcedid_info()`. `lis_result_sourcedid_info()` has no test coverage:

```shellsession
$ make backend-tests coverage | ag validation/_lis_result_sourcedid
lms/validation/_lis_result_sourcedid.py                           13      2    85%   44-46
```

I recommend running `make coverage`, you'd be surprised how often it shows up bugs

Reply:
Thank you for this. Added and unsurprisingly-exposed bug fixed.